sound/soc/sof/ipc4-pcm.c

```diff
@@ -495,17 +496,36 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 				use_chain_dma = true;
 		}
 	}
+
+	fe_rate = params_rate(params);
+
 	/*
-	 * Always set BE format to 32-bits for both playback and capture, except if chained
-	 * DMA is in use. In chained DMA case the back-end and front-end format must match.
+	 * Limit sampling rate if copier has a format that matches
+	 * the rate.
 	 */
 	if (!use_chain_dma) {
-		snd_mask_none(fmt);
-		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+		struct sof_ipc4_available_audio_format *available_fmt = &ipc4_copier->available_fmt;
+		int i;
+
+		for (i = 0; i < available_fmt->num_output_formats; i++) {
+			struct sof_ipc4_audio_format *audio_fmt =
+				&available_fmt->output_pin_fmts[i].audio_fmt;
+
+			if (fe_rate == audio_fmt->sampling_frequency)
+				fe_be_rate_match = true;
+		}
+	}
+
+	if (fe_be_rate_match || use_chain_dma) {
+		/* Limit to common FE/BE sampling rate */
+		rate->min = fe_rate;
+		rate->max = rate->min;
+	} else {
+		rate->min = ipc4_copier->available_fmt.input_pin_fmts->audio_fmt.sampling_frequency;
```

Comment:
@kv2019i the if part is clear, the else not so much. If there's no matching audio format with the fe sampling rate, then why would we choose the first available input arbitrarily? Is this assuming that in this case there will be an SRC in the pipeline that will do the sample rate conversion and that all input formats in topology have the same sampling rate?

Reply:
@ranj063 That's what the existing code does? :) As far as I understand the existing logic (and it's similar in e.g. ipc3-pcm) that we want to limit the sampling rate to that defined in topology. So in vast majority of cases, there's just one rate defined and that is used to set the constraint.

If we have multiple rates in tplg, then we need to pick one and using the FE rate seems to be a good policy here.

If there's no match, then there's either a SRC in the pipeline or the pipeline setup should rightfully fail. We don't really have knowledge in this dai fixup, which rate should we constrain (either multiple options or one option that does not match FE). I remember trying to just not limit the rate here if no match is found with FE, but resulted in many failures. I do think we need to limit the rate somehow here based on topology information.

Reply:
Seems the else-path has limited function as we are checking the formats when we enable the path, so doing the limitation in dai_fixup is maybe not needed.